### ISSUE TITLE
[no-unused-vars] Should not suggest dropping args from overrides

### ISSUE THREAD

**Issue by conartist6** [typescript-eslint/typescript-eslint#4737]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": ["error"]
  }
}
```

```TS
class Base {
  method(arg: number) {
     return arg;
  }
}

class Extended extends Base {
  method(arg: number) { // lint error: unused
     return 'extendedValue';
  }
}
```

**Expected Result**
No lint error. I should define this arg as it is relevant to extending classes

**Actual Result**
Lint error that must be suppressed, and no relevant suppression mechanism exists

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `X.Y.Z` |
| `@typescript-eslint/parser`        | `X.Y.Z` |
| `TypeScript`                       | `X.Y.Z` |
| `ESLint`                           | `X.Y.Z` |
| `node`                             | `X.Y.Z` |

**Comment by bradzacher** [comment]
> I should define this arg as it is relevant to extending classes

No, it's not necessary.
There's no requirement for a subclass to declare the same number of arguments for an overloaded function.
You can delete that argument and no TS error will be reported.

The variable declaration is unnecessary and unused, this this is a correct report.

**Comment by JoshuaKGoldberg** [comment]
> You can delete that argument and no TS error will be reported.

Err, I get an error in the [TypeScript playground](https://www.typescriptlang.org/play?strictFunctionTypes=false&ssl=11&ssc=1&pln=11&pc=2#code/MYGwhgzhAEBCkFNoG8CwAoa0C2CAuAFgPYAmAFGAE4DmAXNAHYCu2ARgpQJQoZZ+X4mlBtCrUA3L2gBfDLPQZQkGAFEAHngQMSCEtAQatJGPAhI0mHPmLluyaFP6Dh0AOQHN23QDUwIJgiukpby0kA):

```
Property 'method' in type 'Extended' is not assignable to the same property in base type 'Base'.
  Type '() => string' is not assignable to type '(arg: number) => number'.
    Type 'string' is not assignable to type 'number'.(
```

Related issues:

* #586: asks for allowing base class methods to have unused parameters, whereas this issue asks for derived class methods
* #1321: this issue is a duplicate of #1321.
* #3759: good further discussion of why this pattern isn't something we consider to be a priority to support

If we did fix for this issue, we would probably want this to be an opt-in option. Seeing only parameters on the methods in a derived class mysteriously not be flagged as unused would be confusing.

However, fixing this issue would require using the TypeScript type checker to know that the parameters match up to a used parameter in the base class. `@typescript-eslint/no-unused-vars` doesn't enable use the TypeScript type checker right now. Doing so would make the rule a good deal slower and necessitate moving it from the `recommended` plugin config to `recommended-requiring-type-checking`.

My recommendation is to not rely on class hierarchies in cases where derived classes don't interact with method parameters the same way as their base classes. 😄

**Comment by bradzacher** [comment]
@JoshuaKGoldberg that TS error is because the example code is bad - the base method returns a number, but the overload returns a string!
The parameter's existence doesn't impact the error.

**Comment by conartist6** [comment]
My thinking is that even if TS knows how to skip a level the hierarchy it's a matter of stylistic choice to write out the full parameters since it's pretty clear what they should be.

**Comment by JoshuaKGoldberg** [comment]
Aha! Thank you, I'll edit my comment.

> stylistic choice

It's not one we would recommend. You're explicitly writing out extra code that you'll need to keep names and types up-to-date in. 👎

**Comment by conartist6** [comment]
Actually I think it might help to have a rule that demands that you explicitly copy the arg types. This [playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBCkFNoG8BQ1oFsEBcAWA9gCYAUYATgOYBc0AdgK6YBGC5AlLYy29AD70GIECnQZo5XA3J1oFSgG4xAX1SrUoSDACyASyJEQSBAA8cCOkRjwISNBmz5iJdqPESpMwcKUZV6zShoADFdOjARU3NLHX1DOzFHQlJ5LiZWDjdxSRxpWXlfaH8gA) shows why.

**Comment by conartist6** [comment]
Rather my point is just that I was wrong -- it is actually more than a semantic difference.

**Comment by bradzacher** [comment]
If you want to have unnecessary arguments - there's an option for that.

https://eslint.org/docs/rules/no-unused-vars#argsignorepattern

---

You're right - this is stylistic! We shy away from stylistic rules because they're complex to build and require many options to satisfy all the styles.

In this case - this isn't a style that's widely used because it introduces unnecessary, unused variables. Thus one we won't be looking to support in our general purpose plugin via an option or a new rule.

**Comment by conartist6** [comment]
Did you look at the [playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEBCkFNoG8BQ1oFsEBcAWA9gCYAUYATgOYBc0AdgK6YBGC5AlLYy29AD70GIECnQZo5XA3J1oFSgG4xAX1SrUoSDACyASyJEQSBAA8cCOkRjwISNBmz5iJdqPESpMwcKUZV6zShoADFdOjARU3NLHX1DOzFHQlJ5LiZWDjdxSRxpWXlfaH8gA)?

My point was that omitting the override arg **isn't just style** -- you're changing the type of the class in a way that can damage your ability to extend it further.

**Comment by conartist6** [comment]
It would seem to me to be best if the `no-unused-vars` rule could apply only on arguments to methods which are not overriding anything, but I'd think that implementing that correctly would require a type-aware implementation of the rule. Would such an implementation be welcome? And then perhaps the default config for type-aware users would suppress the base rule and provide its own

**Comment by conartist6** [comment]
For example in the type-aware version of the rule you could have this:

```js
class Base {
  method(n: number) {
    return n;
  }
}

class extends Base {
  method(n: number) { // no error
    return NaN;
  }
}
```
but also
```js
class Base {
  method(n: number) {
    return n;
  }
}

class extends Base {
  method(n: number, i: number) { // unused var i
    return NaN;
  }
}
```

**Comment by conartist6** [comment]
On the other hand, it's not an uncommon to have the equivalent of an abstract method. Abstract methods can be very useful places to attach documentation particularly in domain-driven designs, so I guess people doing that kind of work would probably just want to disable the rule for methods entirely. For example:

```js
class Base {
  doWork(stuff: any) {
    throw new Error('not implemented');
  }
}
```
